copier.go

```diff
@@ -0,0 +1,150 @@
+package tests
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"math/rand"
+	"testing"
+
+	"github.com/google/uuid"
+	. "github.com/smartystreets/goconvey/convey"
+
+	"github.com/beyondstorage/go-storage/v4/pkg/randbytes"
+	"github.com/beyondstorage/go-storage/v4/services"
+	"github.com/beyondstorage/go-storage/v4/types"
+)
+
+func TestCopier(t *testing.T, store types.Storager) {
+	Convey("Given a basic Storager", t, func() {
+		Convey("The Storager should implement Copier", func() {
+			_, ok := store.(types.Copier)
+			So(ok, ShouldBeTrue)
+		})
+
+		Convey("When Copy a file", func() {
+			c, _ := store.(types.Copier)
+
+			size := rand.Int63n(4 * 1024 * 1024)
+			r := io.LimitReader(randbytes.NewRand(), size)
+			rMD5Hex := calculateHashFromReader(r)
+			src := uuid.New().String()
+
+			_, err := store.Write(src, r, size)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				err = store.Delete(src)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			dst := uuid.New().String()
+			err = c.Copy(src, dst)
+
+			defer func() {
+				err = store.Delete(dst)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			Convey("The error should be nil", func() {
+				So(err, ShouldBeNil)
+			})
+
+			Convey("Read should get dst object data without error", func() {
+				var buf bytes.Buffer
+				n, err := store.Read(dst, &buf)
+
+				Convey("The error should be nil", func() {
+					So(err, ShouldBeNil)
+				})
+
+				Convey("The size should be equal", func() {
+					So(n, ShouldEqual, size)
+				})
+
+				Convey("The hash should be equal", func() {
+					bMD5 := md5.Sum(buf.Bytes())
+					bMD5Hex := hex.EncodeToString(bMD5[:])
+					So(rMD5Hex, ShouldEqual, bMD5Hex)
+				})
+			})
+		})
+
+		Convey("When Copy to an existing file", func() {
+			c, _ := store.(types.Copier)
+
+			sSize := rand.Int63n(4 * 1024 * 1024)
+			sr := io.LimitReader(randbytes.NewRand(), sSize)
+			rMD5Hex := calculateHashFromReader(sr)
+			src := uuid.New().String()
+
+			_, err := store.Write(src, sr, sSize)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			dSize := rand.Int63n(4 * 1024 * 1024)
+			dr := io.LimitReader(randbytes.NewRand(), dSize)
+			dst := uuid.New().String()
+
+			_, err = store.Write(dst, dr, dSize)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				sErr := store.Delete(src)
+				dErr := store.Delete(dst)
+				if sErr != nil {
```

Comment:
a bit strange to me. Why do we need to check the err separately?

Reply:
Can we use the same `err` as the return value of deleting src and dst since `Fail` called by `Error` will marks the function as having failed but continues execution?

Reply:
LGTM.